pkg/storegateway/series_refs.go

```diff
@@ -35,18 +49,51 @@ type seriesChunkRefsIterator interface {
 type seriesChunkRefsSet struct {
 	// series sorted by labels.
 	series []seriesChunkRefs
+
+	// releasable holds whether the series slice (but not its content) can be released to a memory pool.
+	releasable bool
 }
 
-func newSeriesChunkRefsSet(capacity int) seriesChunkRefsSet {
+// newSeriesChunkRefsSet creates a new seriesChunkRefsSet with the given capacity.
+// If releasable is true, then a subsequent call release() will put the internal
+// series slices to a memory pool for reusing.
+func newSeriesChunkRefsSet(capacity int, releasable bool) seriesChunkRefsSet {
+	var prealloc []seriesChunkRefs
+
+	// If it's releasable then we try to reuse a slice from the pool.
+	if releasable {
+		if reused := seriesChunkRefsSetPool.Get(); reused != nil {
+			prealloc = *(reused.(*[]seriesChunkRefs))
+		}
+	}
+
+	if prealloc == nil {
+		prealloc = make([]seriesChunkRefs, 0, capacity)
+	}
+
 	return seriesChunkRefsSet{
-		series: make([]seriesChunkRefs, 0, capacity),
+		series:     prealloc,
```

Comment:
here we don't check whether `cap(prealloc) >= capacity`. So we can request a capacity of 10 and only get a slice with a capacity of 8. This will break this code https://github.com/grafana/mimir/blob/69a31f50a5f888e388e0369556c81893f0460f20/pkg/storegateway/series_refs.go#L522-L538


I think it's not a big deal because we use the same batch size everywhere, but it's a subtle bug

Reply:
Oh this an huge bug. Good job spotting it! Will fix.

Reply:
What code does it break here specifically? Since we initialise the length to 0, I'm not sure I understand why capacity matters here.

Reply:
you're right. capacity doesn't make much of a difference here. the slice will be resized, which is just inefficient, not a bug